File: frontend/src/lib/attachmentConstraints.ts
/**
 * Constraints shared by every attachment-upload widget — FSP Attachments
 * tab (per-FSP attachment management) and the Data Submission screen
 * (attachments accompanying an XML/GeoJSON submission). Keep both
 * uploaders in lockstep by importing from this module rather than
 * redeclaring locally.
 *
 * The 50-byte filename cap matches the database column width
 * (FSP_ATTACHMENT.ATTACHMENT_NAME VARCHAR2(50), and the identically
 * named column on STANDARDS_REGIME_ATTACHMENT); over that, the insert
 * blows up mid-upload with ORA-12899.
 *
 * The cap is measured in UTF-8 BYTES, not characters. Oracle VARCHAR2
 * lengths are byte-semantic by default — a char-semantic column would
 * read VARCHAR2(50 CHAR) — so counting `name.length` (UTF-16 code
 * units) let a 50-"character" name occupy 52 bytes, pass this check,
 * and then fail the insert. An en-dash (U+2013) costs 3 bytes and is
 * what Word and Outlook autocorrect produce from a typed hyphen;
 * accented letters cost 2. Keep this in step with the server-side
 * ca.bc.gov.nrs.fsp.api.util.AttachmentConstraints.
 */

/** Allow-list of file extensions, lowercase with leading dot. */
export const ACCEPTED_ATTACHMENT_EXTENSIONS: readonly string[] = [
  '.pdf',
  '.doc',
  '.docx',
];

/** Max filename length in UTF-8 bytes, including the extension. Matches DB column. */
export const MAX_ATTACHMENT_FILENAME_LEN = 50;

/**
 * UTF-8 byte length of a filename — the value the DB column actually
 * bounds. Differs from `name.length` for any non-ASCII character.
 */
export const filenameByteLength = (filename: string): number =>
  new TextEncoder().encode(filename).length;

/** Per-file upload cap, in bytes (50 MB). */
export const MAX_ATTACHMENT_BYTES = 50 * 1024 * 1024;

/** Case-insensitive `endsWith` check against the allow-list. */
export const hasAcceptedAttachmentExtension = (filename: string): boolean => {
  const lower = filename.toLowerCase();
  return ACCEPTED_ATTACHMENT_EXTENSIONS.some((ext) => lower.endsWith(ext));
};

/**
 * Validate one file against all three constraints in order: extension,
 * filename length, size. Returns the first problem's user-facing
 * `{ title, subtitle }` toast payload, or null on success.
 */
export const validateAttachmentFile = (
  file: File,
): { title: string; subtitle: string } | null => {
  if (!hasAcceptedAttachmentExtension(file.name)) {
    return {
      title: 'Unsupported file type',
      subtitle: `Allowed: ${ACCEPTED_ATTACHMENT_EXTENSIONS.join(', ')}.`,
    };
  }
  const nameBytes = filenameByteLength(file.name);
  if (nameBytes > MAX_ATTACHMENT_FILENAME_LEN) {
    // Report the byte count as the count that matters, but only mention
    // the multi-byte rule when it is actually what tipped the file over —
    // otherwise a plainly-too-long ASCII name gets a confusing hint.
    const inflated = nameBytes > file.name.length;
    return {
      title: 'File name too long',
      subtitle:
        `Max ${MAX_ATTACHMENT_FILENAME_LEN} characters including the extension `
        + `(this file counts as ${nameBytes}`
        + (inflated
          ? `, because accented letters and long dashes count as 2-3 each`
          : '')
        + `). Rename and try again.`,
    };
  }
  if (file.size > MAX_ATTACHMENT_BYTES) {
    return {
      title: 'File too large',
      subtitle: `Max 50 MB (${(file.size / 1_048_576).toFixed(1)} MB)`,
    };
  }
  return null;
};

File: backend/src/main/java/ca/bc/gov/nrs/fsp/api/util/AttachmentConstraints.java
package ca.bc.gov.nrs.fsp.api.util;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Locale;

/**
 * Server-side attachment upload constraints, shared by every upload path
 * (FSP attachments, standards-regime attachments, extension supporting
 * documents). Mirrors the frontend {@code @/lib/attachmentConstraints};
 * both sides must agree or the SPA will accept files the API rejects.
 *
 * <p><b>Filename length is measured in UTF-8 BYTES, not characters.</b>
 * The destination columns — {@code FSP_ATTACHMENT.ATTACHMENT_NAME} and
 * {@code STANDARDS_REGIME_ATTACHMENT.ATTACHMENT_NAME} — are both
 * {@code VARCHAR2(50)} with byte semantics (Oracle's default; a
 * char-semantic column would read {@code VARCHAR2(50 CHAR)}). Counting
 * characters instead let a 50-character name occupy 52 bytes and fail the
 * insert with ORA-12899 <em>after</em> passing validation. That is not
 * hypothetical: an en-dash (U+2013, 3 bytes in UTF-8) is what Word and
 * Outlook autocorrect produce from a typed hyphen, and accented letters
 * cost 2 bytes each.
 *
 * <p>UTF-8 is assumed because the Oracle instance is AL32UTF8. On a
 * single-byte database charset this over-counts slightly, which is the
 * safe direction — it rejects early rather than failing at the insert.
 */
public final class AttachmentConstraints {

  private AttachmentConstraints() {}

  /** Max filename length in UTF-8 bytes, matching ATTACHMENT_NAME VARCHAR2(50). */
  public static final int MAX_FILENAME_BYTES = 50;

  /** Per-file upload cap, in bytes (50 MB). */
  public static final long MAX_ATTACHMENT_BYTES = 50L * 1024 * 1024;

  /** Allow-list of file extensions, lowercase with leading dot. */
  public static final List<String> ACCEPTED_EXTENSIONS = List.of(".pdf", ".doc", ".docx");

  /** UTF-8 byte length of a filename — the value the DB column actually bounds. */
  public static int filenameByteLength(String fileName) {
    return fileName == null ? 0 : fileName.getBytes(StandardCharsets.UTF_8).length;
  }

  /**
   * Validate one upload against extension, filename length, then size —
   * the same order as the frontend so the messages line up. Throws
   * {@link IllegalArgumentException} (mapped to a 400 with a clean
   * message) on the first violation.
   */
  public static void validate(String fileName, long size) {
    String name = fileName == null ? "" : fileName;
    String lower = name.toLowerCase(Locale.ROOT);

    if (ACCEPTED_EXTENSIONS.stream().noneMatch(lower::endsWith)) {
      throw new IllegalArgumentException(
          "File type not supported. Upload a .pdf, .doc, or .docx file.");
    }

    int bytes = filenameByteLength(name);
    if (bytes > MAX_FILENAME_BYTES) {
      // Report the character count (what the user sees) alongside the
      // limit. A name that is visibly "short enough" but still rejected
      // is otherwise baffling — see the note above on en-dashes.
      throw new IllegalArgumentException(
          "File name too long. Use " + MAX_FILENAME_BYTES
              + " characters or fewer, including the extension. Accented letters "
              + "and long dashes count as 2-3 characters each, so this name "
              + "(" + name.length() + " characters) counts as " + bytes + ".");
    }

    if (size > MAX_ATTACHMENT_BYTES) {
      throw new IllegalArgumentException(
          "File exceeds size limit. Max file size is 50 MB. "
              + "Select a smaller file and try again.");
    }
  }
}
